mlflow/utils/async_logging/async_logging_queue.py

```diff
@@ -0,0 +1,161 @@
+"""
+Defines an AsyncLoggingQueue that provides async fashion logging of metrics/tags/params using
+ queue based approach.
+"""
+
+import atexit
+import logging
+import threading
+from concurrent.futures import ThreadPoolExecutor
+from queue import Empty, Queue
+
+from mlflow.entities.metric import Metric
+from mlflow.entities.param import Param
+from mlflow.entities.run_tag import RunTag
+from mlflow.utils.async_logging.run_batch import RunBatch
+from mlflow.utils.async_logging.run_operations import RunOperations
+
+_logger = logging.getLogger(__name__)
+
+
+class AsyncLoggingQueue:
+    """
+    This is a queue based run data processor that queues incoming batches and processes them using
+    single worker thread.
+    """
+
+    def __init__(self, logging_func: callable([str, [Metric], [Param], [RunTag]])) -> None:
+        """
+        Initializes an AsyncLoggingQueue object.
+
+        Args:
+            logging_func: A callable function that takes in four arguments: a string
+                representing the run_id, a list of Metric objects,
+                a list of Param objects, and a list of RunTag objects.
+        """
+        self._queue = Queue()  # Dict[str, Queue]
+        self._logging_func = logging_func
+        self._continue_to_log_data = threading.Event()
+        # Keeping max_workers=1 so that there are no two threads
+        self._BATCH_LOGGING_THREADPOOL = ThreadPoolExecutor(max_workers=1)
+
+        self._BATCH_STATUS_CHECK_THREADPOOL = ThreadPoolExecutor(max_workers=10)
+
+        self._run_data_logging_thread = self._BATCH_LOGGING_THREADPOOL.submit(
+            self._logging_loop
+        )  # concurrent.futures.Future[self._logging_loop]
+
+        atexit.register(self._at_exit_callback)
+
+    def _at_exit_callback(self) -> None:
+        """
+        Callback function to be executed when the program is exiting.
+        Stops the data processing thread and waits for the
+        queue to be drained.
+        Finally, shuts down the thread pools used for data logging and batch processing status
+          check.
+        """
+        try:
+            # Stop the data processing thread
+            self._continue_to_log_data.set()
+            # Waits till queue is drained.
+            self._run_data_logging_thread.result()
+            self._BATCH_LOGGING_THREADPOOL.shutdown(wait=False)
+            self._BATCH_STATUS_CHECK_THREADPOOL.shutdown(wait=False)
+        except Exception as e:
+            _logger.error(f"Encountered error while trying to finish logging: {e}")
+
+    def _logging_loop(self) -> None:
+        """
+        Continuously logs run data until `self._continue_to_process_data` is set to False.
+        If an exception occurs during logging, a `MlflowException` is raised.
+        """
+        try:
+            while not self._continue_to_log_data.is_set():
+                self._log_run_data()
+        except Exception as e:
+            from mlflow.exceptions import MlflowException
```

Comment:
this can be moved to the top level.

Reply:
It causes module load exception, hence importing it here.

Reply:
What's the exception? This shouldn't cause a module load exception